### PR TITLE
[MI-2098]: Refactor phase-1 code base for final release

### PR DESCRIPTION
**Pull request by avas27JTG**
Refactored phase-1 code base for final release

### CODE REVIEW

File: server/plugin/client.go [hunk 1]
```diff
@@ -192,6 +192,15 @@ func (c *client) call(basePath, method, path, contentType string, mattermostUser
 		path = baseURL.String() + path
 	}
 
+	// Check refresh token only for APIs other than OAuth
+	if basePath != constants.BaseOauthURL {
+		if IsAccessTokenExpired, refreshToken := c.plugin.IsAccessTokenExpired(mattermostUserID); IsAccessTokenExpired {
```
Comment (ayusht2810) [inline server/plugin/client.go:197]: `IsAccessTokenExpired` is this corrected in other PR?

File: webapp/src/containers/modals/SubscribeModal/index.tsx [hunk 2]
```diff
@@ -124,7 +126,7 @@ const SubscribeModal = () => {
         }
     };
 
-    // Observe for the change in redux state after API call and do the required actions
+    // Observe for the change in redux state after API call to create subscription and do the required actions
```
Comment (ayusht2810) [inline webapp/src/containers/modals/SubscribeModal/index.tsx:129]:
```suggestion
    // Observe for the change in redux state after the API call to create a subscription and do the required actions
```

File: webapp/src/hooks/useForm.ts [hunk 3]
```diff
@@ -52,8 +52,8 @@ const getValidationErrorMessage = (
     }
 };
 
-// Generic hook to handle form fields
-function useForm(initialFormFields: Record<FormFieldNames, ModalFormFieldConfig>) {
+// Genric hook to handle form fields
```
Comment (ayusht2810) [inline webapp/src/hooks/useForm.ts:55]:
```suggestion
// Generic hook to handle form fields
```

**Review by ayusht2810 — APPROVED** [review]

File: server/plugin/client.go (hunk 1 above)
Comment (avas27JTG) [reply server/plugin/client.go:196]: yes this fixed in master

File: server/store/subscriptions.go [hunk 4]
```diff
@@ -106,6 +106,7 @@ func (s *Store) GetAllSubscriptions(userID string) ([]*serializers.SubscriptionD
 
 	var subscriptionList []*serializers.SubscriptionDetails
 	for _, subscription := range subscriptions.ByMattermostUserID[userID] {
+		subscription := subscription
```
Comment (manojmalik20) [inline server/store/subscriptions.go:98]: You should also add a comment above this line specifying why you are doing this.

File: webapp/src/app.tsx [hunk 5]
```diff
@@ -61,14 +59,19 @@ const App = (): JSX.Element => {
         getSubscribeModalState(usePlugin.state).visibility,
     ]);
 
-    // Fetch the list of linked projects list
+    // Fetch the list of linked project list
```
Comment (manojmalik20) [inline webapp/src/app.tsx:62]:
```suggestion
    // Fetch the list of linked projects
```

File: webapp/src/components/card/index.tsx [hunk 6]
```diff
@@ -18,61 +18,59 @@ type CardProps = {
...
-                )) }
+                        <i className='fa fa-trash'/>
+                    </button>
+                )}
             </div>
         </div>
-    );
-};
+        <div className='card-body'>
+
```
Comment (manojmalik20) [inline webapp/src/components/card/index.tsx:48]:
```suggestion
```

File: webapp/src/components/dropdown/index.tsx [hunk 7]
```diff
@@ -16,7 +16,7 @@ type DropdownProps = {
     error?: boolean | string;
 }
 
-const Dropdown = ({value, placeholder, options, onChange, customOption, loadingOptions, disabled = false, error = '', required}: DropdownProps): JSX.Element => {
+const Dropdown = ({value, placeholder, options, onChange, customOption, loadingOptions, disabled, error, required}: DropdownProps): JSX.Element => {
```
Comment (manojmalik20) [inline webapp/src/components/dropdown/index.tsx:19]: Why this change?

File: webapp/src/containers/Rhs/projectDetails/index.tsx [hunk 8]
```diff
@@ -102,7 +113,7 @@ const ProjectDetails = (projectDetails: ProjectDetails) => {
                 isOpen={showProjectConfirmationModal}
                 onHide={() => setShowProjectConfirmationModal(false)}
                 onConfirm={handleConfirmUnlinkProject}
-                isLoading={usePlugin.getApiState(plugin_constants.pluginApiServiceConfigs.unlinkProject.apiServiceName, projectDetails).isLoading}
+                isLoading={getApiState(plugin_constants.pluginApiServiceConfigs.unlinkProject.apiServiceName, projectDetails).isLoading}
```
Comment (manojmalik20) [inline webapp/src/containers/Rhs/projectDetails/index.tsx:116]: You should create variables for this above the return statement and use that here.

File: webapp/src/containers/Rhs/projectDetails/index.tsx [hunk 9]
```diff
@@ -111,9 +122,9 @@ const ProjectDetails = (projectDetails: ProjectDetails) => {
                 isOpen={showSubscriptionConfirmationModal}
                 onHide={() => setShowSubscriptionConfirmationModal(false)}
                 onConfirm={handleConfirmDeleteSubscription}
-                isLoading={usePlugin.getApiState(plugin_constants.pluginApiServiceConfigs.deleteSubscription.apiServiceName, subscriptionToBeDeleted).isLoading}
+                isLoading={getApiState(plugin_constants.pluginApiServiceConfigs.deleteSubscription.apiServiceName, subscriptionToBeDeleted).isLoading}
                 confirmBtnText='Delete'
-                description='Are you sure you want to delete this subscription?'
+                description='Are you sure you want to delete this subscription ?'
```
Comment (manojmalik20) [inline webapp/src/containers/Rhs/projectDetails/index.tsx:127]: Why this change?

File: webapp/src/containers/modals/LinkModal/index.tsx [hunk 10]
```diff
@@ -0,0 +1,96 @@
+import React, {useEffect} from 'react';
+import {useDispatch} from 'react-redux';
+
```
Comment (manojmalik20) [inline webapp/src/containers/modals/LinkModal/index.tsx:3]: We should remove the empty folders present because of `.gitkeep`. For eg- `containers/action_buttons/.gitkeep`.

File: webapp/src/containers/modals/LinkModal/index.tsx [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+import {getLinkModalState} from 'selectors';
+
+import usePluginApi from 'hooks/usePluginApi';
+import useForm from 'hooks/useForm';
+import useApiRequestCompletionState from 'hooks/useApiRequestCompletionState';
+
+const LinkModal = () => {
+    const {linkProjectModal} = plugin_constants.form;
+
+    // Hooks
```
Comment (manojmalik20) [inline webapp/src/containers/modals/LinkModal/index.tsx:19]:
```suggestion
```

File: webapp/src/containers/modals/SubscribeModal/index.tsx [hunk 12]
```diff
@@ -24,6 +24,8 @@ import Utils from 'utils';
 import './styles.scss';
 
 const SubscribeModal = () => {
+    const {subscriptionModal} = plugin_constants.form;
```
Comment (manojmalik20) [inline webapp/src/containers/modals/SubscribeModal/index.tsx:27]:
```suggestion
    const {subscriptionModal as subscriptionModalFields} = plugin_constants.form;
```

File: webapp/src/containers/modals/TaskModal/index.tsx [hunk 13]
```diff
@@ -89,6 +92,7 @@ const TaskModal = () => {
         }
     };
 
+    // Observe for the change in redux state after API call to create task and do the required actions
```
Comment (manojmalik20) [inline webapp/src/containers/modals/TaskModal/index.tsx:95]:
```suggestion
    // Observe the change in redux state after the API call to create task and do the required actions
```

File: webapp/src/hooks/useForm.ts [hunk 14]
```diff
@@ -75,12 +75,12 @@ function useForm(initialFormFields: Record<FormFieldNames, ModalFormFieldConfig>
     const isErrorInFormValidation = (): boolean => {
         let errorFields = {};
         Object.keys(initialFormFields).forEach((field) => {
-            if (initialFormFields[field as FormFieldNames].validations) {
-                Object.keys(initialFormFields[field as FormFieldNames].validations ?? '').forEach((validation) => {
+            if (initialFormFields[field as FormFieldNames]?.validations) {
+                Object.keys(initialFormFields[field as FormFieldNames]?.validations ?? '').forEach((validation) => {
```
Comment (manojmalik20) [inline webapp/src/hooks/useForm.ts:79]:
```suggestion
                Object.keys(initialFormFields[field as FormFieldNames].validations).forEach((validation) => {
```
Because it is already being checked in the above if condition.

File: webapp/src/hooks/useForm.ts [hunk 15]
```diff
@@ -108,7 +108,7 @@ function useForm(initialFormFields: Record<FormFieldNames, ModalFormFieldConfig>
     };
 
     // Set value for a specific form field
-    const setSpecificFieldValue = (modifiedFormFields: Partial<Record<FormFieldNames, string>>) => {
+    const setSpecificFieldValue = (modifiedFormFields:Partial<Record<FormFieldNames, string>>) => {
```
Comment (manojmalik20) [inline webapp/src/hooks/useForm.ts:111]: Revert this change. Fix lint.

File: webapp/src/selectors/index.tsx [hunk 16]
```diff
@@ -11,15 +11,15 @@ export const getGlobalModalState = (state: ReduxState): GlobalModalState => {
 };
 
 export const getLinkModalState = (state: ReduxState): LinkProjectModalState => {
```
Comment (manojmalik20) [inline webapp/src/selectors/index.tsx:13]: The type of state should be `GlobalState`.

File: webapp/src/hooks/useForm.ts (hunk 15 above)
Comment (avas27JTG) [reply webapp/src/hooks/useForm.ts:111]: reverted these type of change may came due to sync, also eslint is not giving any error for this

File: webapp/src/hooks/useForm.ts (hunk 14 above)
Comment (avas27JTG) [reply webapp/src/hooks/useForm.ts:79]: yes but TS is giving error as `Object is possibly 'undefined'`

File: webapp/src/containers/modals/SubscribeModal/index.tsx (hunk 12 above)
Comment (avas27JTG) [reply webapp/src/containers/modals/SubscribeModal/index.tsx:27]: we cannot use `as` while destructuring
however, we don't need to mention this explicitly as our global type in the `useForm` covers all

File: webapp/src/containers/modals/LinkModal/index.tsx (hunk 11 above)
Comment (avas27JTG) [reply webapp/src/containers/modals/LinkModal/index.tsx:19]: we are writing types for each block as hooks, state variables, etc
why should we remove it?

File: webapp/src/containers/Rhs/projectDetails/index.tsx (hunk 9 above)
Comment (avas27JTG) [reply webapp/src/containers/Rhs/projectDetails/index.tsx:160]: this may have occurred due to syncing branches

File: server/plugin/client.go [hunk 17]
```diff
@@ -201,6 +201,15 @@ func (c *client) Call(basePath, method, path, contentType string, mattermostUser
...
```
[the remark this answers is not in the document]
Comment (avas27JTG) [reply server/plugin/client.go:206]: earlier it was small but later Ayush requested to convert to cap as he need to import and mock for test cases

File: webapp/src/selectors/index.tsx (hunk 16 above)
Comment (avas27JTG) [reply webapp/src/selectors/index.tsx:13]: `GlobalState` does not contain our plugin state `'plugins-mattermost-plugin-azure-devops'` so we have extended the `GlobalState` in a custom interface as `ReduxState` where we have Global state along with our plugin state

File: webapp/src/containers/modals/LinkModal/index.tsx (hunk 11 above)
Comment (manojmalik20) [reply webapp/src/containers/modals/LinkModal/index.tsx:19]: It just doesn't seem useful. You can keep it if you want.

File: webapp/src/containers/modals/SubscribeModal/index.tsx (hunk 12 above)
Comment (manojmalik20) [reply webapp/src/containers/modals/SubscribeModal/index.tsx:27]: I was suggesting to change the variable name to make it more clear.
```suggestion
    const {subscriptionModal: subscriptionModalFields} = plugin_constants.form;
```

File: server/plugin/client.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (manojmalik20) [reply server/plugin/client.go:206]: I don't understand. This variable's scope will only be inside the `if` block. It will not be accessible outside that block.

**Review by manojmalik20 — APPROVED** [review]

File: server/plugin/client.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (avas27JTG) [reply server/plugin/client.go:206]: oh my bad I thought you are saying for everywhere even here `c.plugin.IsAccessTokenExpired(mattermostUserID)`
